### PR TITLE
Linked Time: Adds Fob dragging feature to histogram.

### PR DESCRIPTION
**Pull request by JamesHollyer**
This change uses mouse events to implement a dragging feature on the linked time fob. This allows users to select and drag the fob to change the selected time. This implementation jumps the fob to the next step with data in that histogram which makes it easier for users to see the actual step data.

### CODE REVIEW

File: tensorboard/webapp/widgets/linked_time_fob/linked_time_fob_controller_component.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+} from '@angular/core';
+import {LinkedTime} from '../../metrics/types';
+import {ScaleLinear, ScaleTime} from '../../third_party/d3';
+
+export enum AxisDirection {
+  horizontal = 0,
+  vertical,
+}
+
+export enum DraggingFob {
```
Comment (japie1235813) [inline tensorboard/webapp/widgets/linked_time_fob/linked_time_fob_controller_component.ts:33]: The "Dragging" wording was confusing to me. I think this is used to distinguish which fob the user is interacting with. Could we get ride of the Dragging part and just `Fobs`?

File: tensorboard/webapp/widgets/linked_time_fob/linked_time_fob_controller_component.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  none = 0,
+  startFob,
+  endFob,
```
Comment (japie1235813) [inline tensorboard/webapp/widgets/linked_time_fob/linked_time_fob_controller_component.ts:36]: I think for enum it is better to be all upper case?

File: tensorboard/webapp/widgets/linked_time_fob/linked_time_fob_controller_component.ts (hunk 1 above)
Comment (japie1235813) [reply tensorboard/webapp/widgets/linked_time_fob/linked_time_fob_controller_component.ts:33]: then we can also drop Fob inside
```
enum Fobs {
NONE,
START,
END,
}
```

**Review by japie1235813 — COMMENTED** [review]
thanks for making the changes! I have not finished my review. I will have another run of review ideally tomorrow

File: tensorboard/webapp/widgets/linked_time_fob/linked_time_fob_test.ts [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+    newLinkedTimeActionSpy = jasmine.createSpy();
+    fixture.componentInstance.onSelectedTimeChangedMock =
+      newLinkedTimeActionSpy;
+    newLinkedTimeActionSpy.and.stub();
+
+    return fixture;
+  }
+
+  it('moves the start fob down to mouse when mouse is dragging down and is below fob', () => {
```
Comment (japie1235813) [inline tensorboard/webapp/widgets/linked_time_fob/linked_time_fob_test.ts:102]: Does this mean the mouse is not dragging the fob "in a correct way"? What looks to me is, the mouse is below the fob a bit but still touches the fob and the mouse is moving up. This should also move the fob. If not, could you explain more on what this test is about. Under what scenario it happens?

**Review by JamesHollyer — COMMENTED** [review]
Thanks Jie-Wei! I made some changes and tried explaining some things. Let me know what you think.

File: tensorboard/webapp/widgets/linked_time_fob/linked_time_fob_test.ts [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+  @ViewChild('FobController')
+  fobController!: LinkedTimeFobControllerComponent;
+
+  steps!: number[];
+  axisDirection!: AxisDirection;
+  linkedTime!: LinkedTime;
+  temporalScale!: TemporalScale;
+
+  onSelectedTimeChangedMock(newLinkedTime: LinkedTime) {
+    this.linkedTime = newLinkedTime;
```
[the remark this answers is not in the document]
Comment (japie1235813) [reply tensorboard/webapp/widgets/linked_time_fob/linked_time_fob_test.ts:52]: Please remove the implementation as it is not needed.
`onSelectTimeChanged: (linkedTime: LinkedTime) => void`

Could you add `@Input()` to the params as the example here? 
https://cs.opensource.google/tensorflow/tensorboard/+/master:tensorboard/webapp/widgets/histogram/histogram_test.ts;l=85-90?q=%22class%20TestableComponent%20%7B%22&ss=tensorflow%2Ftensorboard

**Review by japie1235813 — COMMENTED** [review]
This is looking good to me.

Could you check again the `onSelectTimeChanged` implementation of TestableComponent?

**Review by JamesHollyer — COMMENTED** [review]
Thanks for the Review!

**Review by japie1235813 — APPROVED** [review]

**Pull request merged**